lightwood/mixers/bayesian_nn/bayesian_nn.py

```diff
@@ -0,0 +1,307 @@
+        # sample a regressor (which also samples w and b)
+        lifted_reg_model = lifted_module()
+
+        lhat = lifted_reg_model(input_data)
+
+        pyro.sample("obs", pyro.distributions.Categorical(logits=lhat), obs=output_data)
+
+    def pyro_guide(self, input_data, output_data):
+        # First layer weight distribution priors
+        inw_mu = torch.randn_like(self.net.net[0].weight, device=self.net.device)
+        inw_sigma = torch.randn_like(self.net.net[0].weight, device=self.net.device)
+        inw_mu_param = pyro.param("inw_mu", inw_mu)
+        inw_sigma_param = self.softplus(pyro.param("inw_sigma", inw_sigma))
+        inw_prior = pyro.distributions.Normal(loc=inw_mu_param, scale=inw_sigma_param)
+
+        # First layer bias distribution priors
+        inb_mu = torch.randn_like(self.net.net[0].bias, device=self.net.device)
+        inb_sigma = torch.randn_like(self.net.net[0].bias, device=self.net.device)
+        inb_mu_param = pyro.param("inb_mu", inb_mu)
+        inb_sigma_param = self.softplus(pyro.param("inb_sigma", inb_sigma))
+        inb_prior = pyro.distributions.Normal(loc=inb_mu_param, scale=inb_sigma_param)
+
+        # Output layer weight distribution priors
+        outw_mu = torch.randn_like(self.net.net[-1].weight, device=self.net.device)
+        outw_sigma = torch.randn_like(self.net.net[-1].weight, device=self.net.device)
+        outw_mu_param = pyro.param("outw_mu", outw_mu)
+        outw_sigma_param = self.softplus(pyro.param("outw_sigma", outw_sigma))
+        outw_prior = pyro.distributions.Normal(loc=outw_mu_param, scale=outw_sigma_param).independent(1)
+
+        # Output layer bias distribution priors
+        outb_mu = torch.randn_like(self.net.net[-1].bias, device=self.net.device)
+        outb_sigma = torch.randn_like(self.net.net[-1].bias, device=self.net.device)
+        outb_mu_param = pyro.param("outb_mu", outb_mu)
+        outb_sigma_param = self.softplus(pyro.param("outb_sigma", outb_sigma))
+
+        outb_prior = pyro.distributions.Normal(loc=outb_mu_param, scale=outb_sigma_param)
+        priors = {'net[0].weight': inw_prior, 'net[0].bias': inb_prior, 'net[-1].weight': outw_prior, 'net[-1].bias': outb_prior}
+
+        lifted_module = pyro.random_module("module", self.net, priors)
+
+        return lifted_module()
+
+    def predict(self, when_data_source, include_encoded_predictions = False):
+        """
+        :param when_data_source:
+        :return:
+        """
+
+        when_data_source.transformer = self.transformer
+        when_data_source.encoders = self.encoders
+        data_loader = DataLoader(when_data_source, batch_size=len(when_data_source), shuffle=False, num_workers=0)
+
+        self.net.eval()
+        data = next(iter(data_loader))
+        inputs, _ = data
+        inputs = inputs.to(self.net.device)
+
+        sampled_models = [self.pyro_guide(None, None) for _ in range(len(inputs))]
```

Comment:
i dont understand, this should not be a function of the len of inputs. but rather function of a constant NUMBER_OF_MODELS to generate a distribution from (lets start with 100)

Reply:
@George3d6 ^

Reply:
Yes... I'm not sure how that slipped in there, and the length of the input was rather small so I didn't catch it